tutorials/haskell/generics/tutorial.md

```diff
@@ -0,0 +1,584 @@
+-- At the same time if we wanted to derive ‘Functor’, we would work with
+-- ‘Maybe’ of kind * -> *. This means that there are actually two different
+-- possible representations depending on our aim. This is addressed with two
+-- different generics type classes, as we will see later.
+
+-- For kind *, things like ‘Show’:
+
+-- type: (U1 :+: ?) p
+```
+
+Oh, how to represent `Just a`? We need to way to allow it have an argument!
+Let's add the following:
+
+```haskell
+data Rec0 c p = Rec0 { unRec0 :: c }
+```
+
+`Rec` part in the type's name hints that this thing may be possibly
+recursive.
+
+But I'm lying to you, due to some historical stuff, it's defined a bit
+differently:
+
+```haskell
+type Rec0 = K1 R
+
+newtype K1 i c p = K1 { unK1 :: c } -- c is the value, ‘a’ in ‘Maybe a’
+--         ^   ^
+--         |   |
+--         |   +-------- dummy p
+--         |
+-- type-level tag, R or P
+```
+
+You see this `R` type-level tag? There used to be another one, `P` and this
+`type Par0 = K1 P` type synonym, but forget it, it's deprecated. Bottom
+line: `Rec0` is used for data constuctor arguments (fields), that **are
+not** `p` parameter.
+
+With `Rec0`, we can finally build representation of `Maybe a`:
+
+```haskell
+-- This is type of our representation: (U1 :+: Rec0 a) p
+
+-- Examples of values for ‘Maybe Int’:
+
+-- Nothing => L1 U1
+-- Just 5  => R1 (K1 5) -- remember where K1 comes from?
+--            ^
+--            |
+--            +--- L1 and R1 are from our representation of sum types
+```
+
+Let's derive a different representation that works with `* -> *` kinds:
+
+```haskell
+-- Type of our representation: (U1 :+: ?) p
+```
+
+We need a way to tell if we have an argument of `p` type (like `a` in
+`Functor f => f a`) or some other type, if we're to write a generic
+`Functor` function. For this reason the generics extension uses `Par1 p`:
+
+```haskell
+newtype Par1 p = Par1 { unPar1 :: p } -- “par” stands for “parameter”
+```
+
+`Par1` is used to mark occurrences of `p`. Our representation thus becomes:
+
+```haskell
+-- Type of our representation: (U1 :+: Par1) p
+
+-- Examples of values for ‘Maybe Int’:
+
+-- Nothing => L1 U1 -- the same
+-- Just 5  => R1 (Par1 5)
+```
+
+Almost there. The final example is that for lists. Given the standard
+definition of linked list:
+
+```haskell
+data List a = Nil | Cons a (List a)
+```
+
+How do we build its representation (for kind `* -> *`)? The tricky part is,
+of course, `List a`, which is a recursive occurrence of entire functorish
+part with parameter inside it. If we mark occurrences of parameter by
+wrapping them with `Par1`, then why not mark this thing with something too?
+`Rec1` shall it be.
+
+```haskell
+-- Type of representation: (U1 :+: (Par1 :*: Rec1 List))
+
+-- Examples of values for ‘List Int’:
+
+-- Nil                 => L1 U1
+-- Cons 5 Nil          => R1 (Par1 5 :*: Rec1 Nil)
+-- Cons 5 (Cons 4 Nil) => R1 (Par1 5 :*: Rec1 (Cons 4 Nil))
+```
+
```

Comment:
My confusion about the `p` parameter from above (line 179) continues here.  What distinguishes parameters of "type p" from other type parameters?
If the following paragraph is left as is (and perhaps with some more explanation about `p` above it could be):
... both the first and the second arguments ...